### ISSUE TITLE
Why do some visualization pictures and result pictures not match?

### ISSUE THREAD

**Issue by shijinghuihub** [jacobgil/pytorch-grad-cam#282]
Hi, Thanks for your great work. I use it to visualize my binary segmentation model.
My outputs are : 
```
c = self.conv0(c1)  # b,1,h,w
z = self.convz(z)   # b,1,h,w
return [c,z]
```
My visualization code is 
```
class SegmentationModelOutputWrapper(torch.nn.Module):
    def __init__(self,cfg): 
        super(SegmentationModelOutputWrapper, self).__init__()
        self.model = M2LP(cfg,"M2LP").eval()
        # self.mask = mask
    def forward(self, x):
        return self.model(x)[0]  
``` 

```
class SemanticSegmentationTarget:
    def __init__(self):
        pass
        # self.category = category
        # self.mask = torch.from_numpy(mask)
        # if torch.cuda.is_available():
        #     self.mask = self.mask.cuda()
        
    def __call__(self, model_output):
        return model_output.sum()

def makedirs(path):
    if os.path.exists(path) == False:
        os.makedirs(path)
```

```
if __name__=='__main__':

    model = SegmentationModelOutputWrapper(cfg) 
    model.cuda()
    i=0
    for _ in tqdm(dataset_list):
        root = os.path.join(dataset_list[i], 'images')
        fileList = os.listdir(root) 
        k=0
        for _ in tqdm(fileList):
            image = Image.open(os.path.join(root,fileList[k]))
            size = 320
            image = image.resize((size, size))
            image = np.array(image)
            rgb_img = np.float32(image) / 255
            input_tensor = preprocess_image(rgb_img,
                                        mean=[0.485, 0.456, 0.406],
                                        std=[0.229, 0.224, 0.225])
            input_tensor = input_tensor.cuda()

            target_layers = [model.model.decoder.conv0]  
            targets1 = [SemanticSegmentationTarget()]
            with GradCAM(model=model,
                    target_layers=target_layers,
                    use_cuda=torch.cuda.is_available()) as cam:
                grayscale_cam = cam(input_tensor=input_tensor,targets=targets1)[0, :]
                cam_image = show_cam_on_image(rgb_img, grayscale_cam, use_rgb=True)
            a = Image.fromarray(cam_image)
            path = os.path.join(save_root,dataset_list[i])
            makedirs(path)
            a.save(os.path.join(path,fileList[k]))
            k = k + 1
        i = i + 1
```

I'm confused why some of the visualizations don't match my predictions

**Comment by jacobgil** [comment]
Hi,
So the model has one output, not two, right ?

The issue here is that then the output is always the score for the second category.
So model_output.sum() in the target function, will always create the CAM for the second category, even if the prediction was less than 0.5.

You can control the category the CAM is created for, by passing the category you want to create the mask for to SemanticSegmentationTarget.
And then do model_output.sum() * (2 * category - 1), so if category==0, it will create negative gradients and the CAM for the first category.

**Comment by ChenHongyu-bo** [comment]
> Hi, So the model has one output, not two, right ?
> 
> The issue here is that then the output is always the score for the second category. So model_output.sum() in the target function, will always create the CAM for the second category, even if the prediction was less than 0.5.
> 
> You can control the category the CAM is created for, by passing the category you want to create the mask for to SemanticSegmentationTarget. And then do model_output.sum() * (2 * category - 1), so if category==0, it will create negative gradients and the CAM for the first category.

Hi, I also encountered the same problem, the visualization results are very different from my prediction results, can you help me check the code? By the way, my network ends up outputting only one class

`modelpath=r'./Epoch_27_Loss_0.059_Acc_0.990_Iou_0.911.pkl'
model = Net()
model.load_state_dict(torch.load(modelpath))
model.eval()

image_path=r'./2_1_0.tif'
img=io.imread(image_path)
# img = np.array(img)
rgb_img = np.float32(img) / 255
input_tensor = preprocess_image(rgb_img,
                                mean=[0.485, 0.456, 0.406],
                                std=[0.229, 0.224, 0.225])

if torch.cuda.is_available():
    model = model.cuda()
    input_tensor = input_tensor.cuda()


res, o1, o2, o3, o4= model(input_tensor)


output  = res .sigmoid().data.cpu().numpy().squeeze()


# plopymaskuint8 = 255 * np.uint8(output  ==1)
plopymaskfloat = np.float32(output>=0.5)


from pytorch_grad_cam import GradCAM

targets = [SemanticSegmentationTarget(0, plopymaskfloat)]
target_layers = [model.msp.merge_feat]
# target_layers = [model.encoder.conv1x1]

with GradCAM(model=model,target_layers=target_layers,use_cuda=torch.cuda.is_available()) as cam:

    grayscale_cam = cam(input_tensor=input_tensor,targets=targets,)[0, :]

    cam_image = show_cam_on_image(rgb_img, grayscale_cam, use_rgb=True)
    Image.fromarray(cam_image)
    plt.imshow(cam_image)
    plt.show()
    # cv.imwrite("123.png",cam_image)
`

**Comment by shijinghuihub** [comment]
> Hi, So the model has one output, not two, right ?
> 
> The issue here is that then the output is always the score for the second category. So model_output.sum() in the target function, will always create the CAM for the second category, even if the prediction was less than 0.5.
> 
> You can control the category the CAM is created for, by passing the category you want to create the mask for to SemanticSegmentationTarget. And then do model_output.sum() * (2 * category - 1), so if category==0, it will create negative gradients and the CAM for the first category.

Sorry, I need to rephrase my problem is that  my prediction result can segment all the foreground, but the cam result loses part of it.
